test/gtest/wat-lexer.cpp

```diff
@@ -101,1446 +82,864 @@ TEST(LexerTest, LexParens) {
-    ASSERT_FALSE(lexer.empty());
-    Token expected{"-9223372036854775808"sv, IntTok{uint64_t(INT64_MIN), Neg}};
-    EXPECT_EQ(*lexer, expected);
-  }
-  {
-    Lexer lexer("-9223372036854775809"sv);
-    ASSERT_FALSE(lexer.empty());
-    Token expected{"-9223372036854775809"sv,
-                   IntTok{uint64_t(INT64_MIN) - 1, Neg}};
-    EXPECT_EQ(*lexer, expected);
-  }
+  EXPECT_EQ(Lexer("0"sv).takeU8(), uint8_t(0));
+  EXPECT_EQ(Lexer("0"sv).takeI8(), uint8_t(0));
+  EXPECT_EQ(Lexer("0"sv).takeI16(), uint16_t(0));
+  EXPECT_EQ(Lexer("0"sv).takeU32(), uint32_t(0));
+  EXPECT_EQ(Lexer("0"sv).takeI32(), uint32_t(0));
+  EXPECT_EQ(Lexer("0"sv).takeU64(), uint64_t(0));
+  EXPECT_EQ(Lexer("0"sv).takeI64(), uint64_t(0));
+
+  EXPECT_FALSE(Lexer("+0"sv).takeU8());
+  EXPECT_EQ(Lexer("+0"sv).takeI8(), uint8_t(0));
+
+  EXPECT_FALSE(Lexer("-0"sv).takeU8());
+  EXPECT_EQ(Lexer("-0"sv).takeI8(), uint8_t(0));
+
+  EXPECT_EQ(Lexer("1"sv).takeU8(), uint8_t(1));
+  EXPECT_EQ(Lexer("1"sv).takeI8(), uint8_t(1));
+
+  EXPECT_FALSE(Lexer("+1"sv).takeU8());
+  EXPECT_EQ(Lexer("+1"sv).takeI8(), uint8_t(1));
+
+  EXPECT_FALSE(Lexer("-1"sv).takeU8());
+  EXPECT_EQ(Lexer("-1").takeI8(), uint8_t(-1));
+
+  EXPECT_EQ(Lexer("0010"sv).takeU8(), uint8_t(10));
+  EXPECT_EQ(Lexer("0010"sv).takeI8(), uint8_t(10));
+
+  EXPECT_FALSE(Lexer("+0010"sv).takeU8());
+  EXPECT_EQ(Lexer("+0010"sv).takeI8(), uint8_t(10));
+
+  EXPECT_FALSE(Lexer("-0010"sv).takeU8());
+  EXPECT_EQ(Lexer("-0010"sv).takeI8(), uint8_t(-10));
+
+  EXPECT_FALSE(Lexer("9999"sv).takeU8());
+  EXPECT_EQ(Lexer("9999"sv).takeI16(), uint16_t(9999));
+  EXPECT_EQ(Lexer("9999"sv).takeU32(), uint32_t(9999));
+  EXPECT_EQ(Lexer("9999"sv).takeI32(), uint32_t(9999));
+
+  EXPECT_FALSE(Lexer("+9999"sv).takeU32());
+  EXPECT_EQ(Lexer("+9999"sv).takeI32(), uint32_t(9999));
+
+  EXPECT_FALSE(Lexer("-9999"sv).takeU32());
+  EXPECT_EQ(Lexer("-9999"sv).takeI32(), uint32_t(-9999));
+
+  EXPECT_EQ(Lexer("12_34"sv).takeU32(), uint32_t(1234));
+  EXPECT_EQ(Lexer("12_34"sv).takeI32(), uint32_t(1234));
+
+  EXPECT_EQ(Lexer("1_2_3_4"sv).takeU32(), uint32_t(1234));
+  EXPECT_EQ(Lexer("1_2_3_4"sv).takeI32(), uint32_t(1234));
+
+  EXPECT_FALSE(Lexer("_1234"sv).takeU32());
+  EXPECT_FALSE(Lexer("_1234"sv).takeI32());
+
+  EXPECT_FALSE(Lexer("1234_"sv).takeU32());
+  EXPECT_FALSE(Lexer("1234_"sv).takeI32());
+
+  EXPECT_FALSE(Lexer("12__34"sv).takeU32());
+  EXPECT_FALSE(Lexer("12__34"sv).takeI32());
+
+  EXPECT_FALSE(Lexer("12cd56"sv).takeU32());
+  EXPECT_FALSE(Lexer("12cd56"sv).takeI32());
+
+  EXPECT_EQ(Lexer("18446744073709551615"sv).takeU64(), uint64_t(-1));
+  EXPECT_EQ(Lexer("18446744073709551615"sv).takeI64(), uint64_t(-1));
+
+  // 64-bit overflow!
+  EXPECT_FALSE(Lexer("18446744073709551616"sv).takeU64());
+  EXPECT_FALSE(Lexer("18446744073709551616"sv).takeI64());
+
+  EXPECT_FALSE(Lexer("+9223372036854775807"sv).takeU64());
+  EXPECT_EQ(Lexer("+9223372036854775807"sv).takeI64(), INT64_MAX);
```

Comment:
I am having trouble mapping these two lines to the old tests. Based on the constant, I looked at

```cpp
  {
    Lexer lexer("+9223372036854775807"sv);
    ASSERT_FALSE(lexer.empty());
    Token expected{"+9223372036854775807"sv, IntTok{INT64_MAX, Pos}};
    EXPECT_EQ(*lexer, expected);
  }
```
But the first part looks different? Before we checked the lexer was not empty, and now we check that `takeU64` is false (why is it false?)

Reply:
`takeU*` can never succeed when the token starts with `+` or `-` because unsigned numbers should not have signs. `take*`, on the other hand, falls back to parsing the number as signed then reinterpreting it to be unsigned, so that succeeds.

Reply:
Thanks, and is that a new test compared to before? Or am I not reading the old code right?

Reply:
I see, thanks!